src/ripple/app/tx/impl/MPTokenAuthorize.cpp

```diff
@@ -46,68 +46,85 @@ MPTokenAuthorize::preflight(PreflightContext const& ctx)
 TER
 MPTokenAuthorize::preclaim(PreclaimContext const& ctx)
 {
-    auto const sleMptIssuance =
-        ctx.view.read(keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
-    if (!sleMptIssuance)
-        return tecOBJECT_NOT_FOUND;
-
     auto const accountID = ctx.tx[sfAccount];
     auto const txFlags = ctx.tx.getFlags();
     auto const holderID = ctx.tx[~sfMPTokenHolder];
 
     if (holderID && !(ctx.view.exists(keylet::account(*holderID))))
         return tecNO_DST;
 
-    std::uint32_t const mptIssuanceFlags = sleMptIssuance->getFieldU32(sfFlags);
-
-    // If tx is submitted by issuer, they would either try to do the following
-    // for allowlisting:
-    // 1. authorize an account
-    // 2. unauthorize an account
+    // if non-issuer account submits this tx, then they are trying either:
+    // 1. Unauthorize/delete MPToken
+    // 2. Use/create MPToken
     //
-    // Note: `accountID` is issuer's account
-    //       `holderID` is holder's account
-    if (accountID == (*sleMptIssuance)[sfIssuer])
+    // Note: `accountID` is holder's account
+    //       `holderID` is NOT used
+    if (!holderID)
     {
-        // If tx is submitted by issuer, it only applies for MPT with
-        // lsfMPTRequireAuth set
-        if (!(mptIssuanceFlags & lsfMPTRequireAuth))
-            return tecNO_AUTH;
+        std::shared_ptr<SLE const> sleMpt = ctx.view.read(
+            keylet::mptoken(ctx.tx[sfMPTokenIssuanceID], accountID));
+
+        // There is an edge case where holder deletes MPT after issuance has
+        // already been destroyed. So we must check for unauthorize before
+        // fetching the MPTIssuance object(since it doesn't exist)
+
+        // if holder wants to delete/unauthorize a mpt
+        if (txFlags & tfMPTUnauthorize)
+        {
+            if (!sleMpt)
+                return tecNO_ENTRY;
 
-        if (!holderID)
+            if ((*sleMpt)[sfMPTAmount] != 0)
+                return tecHAS_OBLIGATIONS;
+
+            return tesSUCCESS;
+        }
+
+        // Now test when the holder wants to hold/create/authorize a new MPT
+        auto const sleMptIssuance =
+            ctx.view.read(keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
+
+        if (!sleMptIssuance)
+            return tecOBJECT_NOT_FOUND;
```

Comment:
Why returning `tecOBJECT_NOT_FOUND` here and `tecNO_ENTRY` above? Shouldn't it be the former in both cases?